Source/RestKit/RestRequest.swift

```diff
@@ -103,17 +103,64 @@ internal struct RestRequest {
         self.request = request
     }
 
-    internal func response(completionHandler: @escaping (Data?, HTTPURLResponse?, Error?) -> Void) {
+    internal func response(
+        parseServiceError: ((HTTPURLResponse?, Data?) -> Error?)? = nil,
+        completionHandler: @escaping (Data?, HTTPURLResponse?, Error?) -> Void)
+    {
         let task = session.dataTask(with: request) { (data, response, error) in
-            completionHandler(data, response as? HTTPURLResponse, error)
+
+            guard error == nil  else {
+                completionHandler(data, response as? HTTPURLResponse, error)
+                return
+            }
+
+            guard let response = response as? HTTPURLResponse else {
+                let error = RestError.noResponse
+                completionHandler(data, nil, error)
```

Comment:
I wonder if we should pull apart the two cases of `response == nil` and `response` is not an `HTTPURLResponse`.  It looks wrong to me to pass a response with an error that says "no response".  I add a `responseNotValid` case to RestError and return this when the response is present by not castable to `HTTPURLResponse`.

Reply:
Ah, gotcha. You're right. We should just leave it as-is.

My concern was that we may write a completion handler that expects a `URLResponse` that is _not_ an `HTTPURLResponse` (for whatever reason). It would silently fail because this block passes `nil`, which might take awhile to find while debugging. (In fact, it wouldn't actually be silent but a failure of `no response`, which could even send us down a completely different debugging path than looking at this block.)

But none of that could actually happen because—as you noted—the `completionHandler` already requires an `HTTPURLResponse`. So my concern would be caught in the type system at compile time instead of a runtime error.